third_party/iap_jwt_verify_nginx/src/module.cc

```diff
@@ -0,0 +1,541 @@
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+// SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "src/module.h"
+
+#include <mutex>
+
+#include "src/iap_jwt_verification.h"
+
+namespace google {
+namespace cloud {
+namespace iap {
+
+// Only trust IAP state for a maximum of five minutes.
+constexpr int MAX_STATE_CACHE_TIME_SEC = 300;  // 5 minutes
+
+// While the theoretically time to cache the JWT verification keys is 1 day,
+// refresh every 12 hours to be on the safe side.
+constexpr int MAX_KEY_CACHE_TIME_SEC = 43200;  // 12 hours
+
+// nginx lowercases headers prior to hashing, so using a lowercase value here
+constexpr char IAP_JWT_HEADER_NAME[] = "x-goog-iap-jwt-assertion";
+
+// A constant used to construct the expected JWT audience.
+constexpr char IAP_JWT_AUD_PROJECTS[] = "/projects/";
+
+// A constant used to construct the expected JWT audience.
+constexpr char IAP_JWT_AUD_APPS[] = "/apps/";
+
+// Used to synchronize state updates.
+ngx_atomic_t iap_state_lock;
+
+// Used to synchronize key updates and access to the main conf's shared key
+// pointer.
+std::mutex key_mutex;
+
+// calculate the hash of an ngx_str_t
+ngx_uint_t hash_ngx_string(ngx_str_t const *const str) {
+  ngx_uint_t hash = 0;
+  for (size_t i = 0; i < str->len; i++) {
+    hash = ngx_hash(hash, str->data[i]);
+  }
+  return hash;
+}
+
+// Extract the X-Goog-Iap-Jwt-Assertion header. Unfortunately, nginx internals
+// limit us to a linear search (without reaching in and modifying the guts,
+// versus just compiling in a new module).
+ngx_str_t *extract_iap_jwt_header(ngx_http_request_t *r) {
+  ngx_list_part_t *part;
+  ngx_table_elt_t *h;
+  ngx_uint_t i;
+
+  part = &r->headers_in.headers.part;
+  h = reinterpret_cast<ngx_table_elt_t *>(part->elts);
+
+  static const ngx_str_t HEADER_NAME = ngx_string(IAP_JWT_HEADER_NAME);
+  for (i = 0;; i++) {
+    if (i >= part->nelts) {
+      if (part->next == nullptr) {
+        break;
+      }
+
+      part = part->next;
+      h = reinterpret_cast<ngx_table_elt_t *>(part->elts);
+      i = 0;
+    }
+
+    if (HEADER_NAME.len != h[i].key.len
+        || ngx_strcasecmp(HEADER_NAME.data, h[i].key.data) != 0) {
+      continue;
+    }
+
+    return &h[i].value;
+  }
+
+  return nullptr;
+}
+
+// IAP JWT verification handler--core business logic lives here and in functions
+// called by this handler.
+ngx_int_t ngx_http_iap_jwt_verification_handler(ngx_http_request_t *r) {
+  ngx_iap_jwt_verify_loc_conf_t *loc_conf =
+      reinterpret_cast<ngx_iap_jwt_verify_loc_conf_t *>(
+          ngx_http_get_module_loc_conf(r, ngx_iap_jwt_verify_module));
+
+  if (loc_conf->iap_jwt_verify == 0) {
+    return NGX_OK;
```

Comment:
Nope, iap_jwt_verify is the flag that indicates whether IAP JWT enforcement should be done for a given location. If it equals zero, enforcement is not being done, which means access is allowed. This handler's job, if called, is to make an access decision--and if no enforcement is being done, then the decision has to be "OK, access allowed".

I'll add a comment to the code.